test/integration_test.go

```diff
@@ -233,56 +247,76 @@ func TestIntegration_ManyVotesScenario(t *testing.T) {
 
 // -----------------------------------------------------------------------------
 // Utility functions
-func newTxManager(signer crypto.Signer, firstNode dVotingCosiDela, timeout time.Duration) txManager {
+func newTxManager(signer crypto.Signer, firstNode dVotingCosiDela,
+	timeout time.Duration, retry int) txManager {
+
+	client := client{
+		srvc: firstNode.GetOrdering(),
+		mgr:  firstNode.GetValidationSrv(),
+	}
+
 	return txManager{
-		m: signed.NewManager(signer, &txClient{}),
-		n: firstNode,
-		t: timeout,
+		m:     signed.NewManager(signer, client),
+		n:     firstNode,
+		t:     timeout,
+		retry: retry,
 	}
 }
 
 type txManager struct {
-	m txn.Manager
-	n dVotingCosiDela
-	t time.Duration
+	m     txn.Manager
+	n     dVotingCosiDela
+	t     time.Duration
+	retry int
 }
 
 func (m txManager) addAndWait(args ...txn.Arg) ([]byte, error) {
-	err := m.m.Sync()
-	if err != nil {
-		return nil, xerrors.Errorf("failed to Sync: %v", err)
-	}
+	for i := 0; i < m.retry; i++ {
+		sentTxn, err := m.m.Make(args...)
+		if err != nil {
+			return nil, xerrors.Errorf("failed to Make: %v", err)
+		}
 
-	sentTxn, err := m.m.Make(args...)
-	if err != nil {
-		return nil, xerrors.Errorf("failed to Make: %v", err)
-	}
-	sentTxnID := sentTxn.GetID()
+		ctx, cancel := context.WithTimeout(context.Background(), m.t)
+		defer cancel()
 
-	err = m.n.GetPool().Add(sentTxn)
-	if err != nil {
-		return nil, xerrors.Errorf("failed to Add: %v", err)
-	}
+		events := m.n.GetOrdering().Watch(ctx)
 
-	ctx, cancel := context.WithTimeout(context.Background(), m.t)
-	defer cancel()
+		err = m.n.GetPool().Add(sentTxn)
+		if err != nil {
+			return nil, xerrors.Errorf("failed to Add: %v", err)
+		}
 
-	events := m.n.GetOrdering().Watch(ctx)
+		sentTxnID := sentTxn.GetID()
 
-	for event := range events {
-		for _, result := range event.Transactions {
-			fetchedTxnID := result.GetTransaction().GetID()
+	events:
+		for event := range events {
+			for _, result := range event.Transactions {
+				fetchedTxnID := result.GetTransaction().GetID()
 
-			if bytes.Equal(sentTxnID, fetchedTxnID) {
-				accepted, status := event.Transactions[0].GetStatus()
+				if bytes.Equal(sentTxnID, fetchedTxnID) {
+					accepted, status := event.Transactions[0].GetStatus()
 
-				if !accepted {
-					return nil, xerrors.Errorf("transaction has not been accepted: %s", status)
-				}
+					if !accepted {
+						if i+1 == m.retry {
+							return nil, xerrors.Errorf("transaction not accepted: %s", status)
+						}
+
+						// let's retry and sync the nonce in case
+						err = m.m.Sync()
+						if err != nil {
+							return nil, xerrors.Errorf("failed to sync: %v", err)
+						}
 
-				return sentTxnID, nil
+						break events
+					}
+
+					return sentTxnID, nil
+				}
```

Comment:
this, on the surface at least, looks like a good candidate for decomposition into separate functions.
In particular I'd consider using a sentinel error instead of a break, I believe this could simplify the code a lot.